Fix mailserver spec (#64)
status-im/specs: status-whisper-mailserver-spec.md

### Archiving messages

> A node which wants to provide a mailserver capability MUST store envelopes incoming in messages packets identify by a code `0x01`. The envelopes can be stored in an arbitrary format, however, they MUST be serialized and deserialized to Whisper envelope format.

Reviewer: ```suggestion
A node which wants to provide mailserver functionality MUST store envelopes from incoming messages packets (Whisper packet-code `0x01`). The envelopes can be stored in an arbitrary format, however, they MUST be serialized and deserialized to Whisper envelope format.
```
I prefer not to use the word capability as that is a term used in rlpx for subprotocols. This could create confusion

> A node which wants to provide mailserver functionality MUST store envelopes from incoming messages packets (Whisper packet-code `0x01`). The envelopes can be stored in an arbitrary format, however, they MUST be serialized and deserialized to Whisper envelope format.

Reviewer 2: ```suggestion
A node which wants to provide mailserver functionality MUST store envelopes from incoming message packets (Whisper packet-code `0x01`). The envelopes can be stored in any format, however they MUST be serialized and deserialized to the Whisper envelope format.
```

> A mailserver SHOULD store envelopes for all topics to be useful by any peer, however, when for personal use case it MAY store only subset of topics.

Reviewer: ```suggestion
A mailserver SHOULD store envelopes for all topics to be generally useful for any peer, however, for specific use cases it MAY store envelopes for a subset of topics.
```

> A mailserver SHOULD store envelopes for all topics to be generally useful for any peer, however, for specific use cases it MAY store envelopes for a subset of topics.

Reviewer 2: ```suggestion
A mailserver SHOULD store envelopes for all topics to be generally useful for any peer, however for specific use cases it MAY store envelopes for a subset of topics.
```

### Requesting messages

> In order to request historic messages, a node MUST send a packet P2P Request (`0x7e`) to its peer providing mailserver capability. This packet requires one argument which MUST be a Whisper envelope.

Reviewer: ```suggestion
In order to request historic messages, a node MUST send a packet P2P Request (`0x7e`) to its peer providing mailserver functionality. This packet requires one argument which MUST be a Whisper envelope.
```

In the Whisper envelope's payload section, there MUST be RLP-encoded information about the details of the request:

```
[ Lower, Upper, Bloom, Limit, Cursor ]
```

`Lower`: 4-byte wide unsigned integer (UNIX time in seconds; oldest requested envelope's creation time)
`Upper`: 4-byte wide unsigned integer (UNIX time in seconds; newest requested envelope's creation time)
`Bloom`: 64-byte wide array of Whisper topics encoded in a bloom filter to filter envelopes
`Limit`: 4-byte wide unsigned integer limiting the number of returned envelopes

> `Cursor`: 32-byte wide array of a cursor returned from the previous request (optional)

Reviewer 3: How is this returned exactly?

What does it correspond to? I suppose an address of some kind that can be used for that specific mailserver to skip previously fetched messages? Does it send messages 'before' this cursor or after? How does it impact Lower and Upper, i.e. does it cap it to be higher than Lower and lower than Higher according to some logic?

Author: It's up to the implementation. Also, it's not possible to query one mailserver and use the obtained cursor to query another one -- I will make it clear. It is so because each mailserver can obtain messages in different order or even may miss some. So it's Whisper property not this spec.

All other questions are up to the implementation due to what I wrote above.

> The envelope MUST be signed with a symmetric key agreed between the requester and Mailserver.

Reviewer: How is this symmetric key agreed up on?

Reviewer: Just realized that this does not need to be part of the spec and can be arranged out of band (layer higher, etc.)
However, should we then not leave the option open to use either a symmetric or asymmetric key? Such as with the regular posting of messages?

Author: Yes, I should make it more generic. It just must be encrypted somehow because each envelope requires that.

Reviewer 3: Seems like a feature enhancement request to spec to me, I suggest we note that it has to be agreed out of band in implementer notes in waku spec

For Status app it is hardcoded, iirc.

Author: It is hardcoded, correct. I will keep it as it is then for now.

### Receiving historic messages

> Historic messages MUST be sent to a peer with a packet with a P2P Message code (`0x7f`) followed by a single Whisper envelope OR an array of Whisper envelopes. A peer receiving historic message MUST handle both cases.

Reviewer: This is a bit problematic. 
1. I believe it goes against the waku/0 or Whisper/6 spec as there the P2P message code is defined with a single envelope as parameter. 
2. In our nim implementation (DSL) it is not supported to do such thing (1 envelope or list of envelopes) .
So I would prefer to have waku/0 altered with either always a list of envelopes (can be 1). Or to add a message packet for multiple envelopes.

Author: > I believe it goes against the waku/0 or Whisper/6 spec as there the P2P message code is defined with a single envelope as parameter.

Correct but in the case when a mailserver returns a lot of messages, sending them one by one adds huge overhead. That's why we batched them.

> In our nim implementation (DSL) it is not supported to do such thing (1 envelope or list of envelopes) .
So I would prefer to have waku/0 altered with either always a list of envelopes (can be 1). Or to add a message packet for multiple envelopes.

I would opt for multiple envelopes then. If one wants to send just one, they can do that as one-item array. The problem is that it makes Waku incompatible with Whisper... Can we solve this in the bridge?

> Historic messages MUST be sent to a peer with as packet with a P2P Message code (`0x7f`) followed by a single Whisper envelope OR an array of Whisper envelopes. A peer receiving historic message MUST handle both cases.

Reviewer 3: The single or array seems a bit suboptimal to me, ideally this should be just array and then it can even be empty, just like Messages packet.

Seems like an oversight in initial EIP627, perhaps this interface is something we can fix for later versions of waku.

Reviewer: Yes, agree. That's why I suggested either changing this packet like you say -> array which can be empty. Or adding a new packet id, which does this, to the waku spec.

Author: [I replied with the same suggestions](https://github.com/status-im/specs/pull/64#discussion_r349266498). Just note that it will be incompatible with EIP-627.

> Historic messages MUST be sent to a peer with a packet with a P2P Message code (`0x7f`) followed by a single Whisper envelope OR an array of Whisper envelopes. A peer receiving historic message MUST handle both cases.

Reviewer 2: ```suggestion
Historic messages MUST be sent to a peer with as packet with a P2P Message code (`0x7f`) followed by a single Whisper envelope OR an array of Whisper envelopes. A peer receiving historic message MUST handle both cases.
```

> In order to accept a P2P Message packet, a node MUST trust a selected Mailserver.

Reviewer 3: Can we be more specific about what we mean by "trust"? I suppose this means making certain envelope invariants / checks optional?

E.g. here is roughly the logic in nim-eth as far as I can tell https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx_protocols/whisper_protocol.nim#L223-L225

Author: Good point. Generally, a trusted peer in Whisper terminology means it's allowed to send p2p messages.
